File: services/bake/app/geom/project.py
"""Local metric frame: WGS84 -> UTM -> centered ENU, plus rotation and crop.

This module is the boundary where degrees stop existing. Everything it returns
is meters in a local ENU frame with the requested center at (0, 0), x east,
y north (03 "Projection and crop"). Web Mercator is never used for geometry.

Conventions
-----------
* UTM zone is derived from the *center* of the request; the whole scene uses
  that one zone even if the crop straddles a zone boundary, so the frame stays
  a single rigid metric frame.
* ``rotation_deg`` is applied as a mathematically positive (counter-clockwise)
  rotation of the geometry about the origin.  Equivalently: the compass bearing
  ``rotation_deg`` (degrees clockwise from true north) ends up pointing at +y
  in the model.  That is why the New York preset is 29 deg: Manhattan's avenue
  grid runs ~29 deg east of north, so bearing 29 -> "up" squares the grid to
  the plate.
* Rotation is a rigid transform, so it is applied to the raw coordinate arrays
  right after projection (one vectorized numpy pass) rather than per geometry.
  Distance-based hygiene (simplify tolerance, centroid dedupe) is rotation
  invariant, so "rotate then crop" per 03 is preserved exactly.
"""
from __future__ import annotations

import math
from typing import Iterable, Sequence

import numpy as np
from pyproj import Transformer
from shapely.geometry import LineString, MultiLineString, MultiPolygon, Polygon, box
from shapely.geometry.base import BaseGeometry

__all__ = [
    "LocalFrame",
    "utm_epsg",
    "utm_zone",
    "crop_square",
    "clip_polygon",
    "clip_line",
    "in_square",
]


def utm_zone(lon: float) -> int:
    """UTM zone number (1..60) for a longitude in degrees."""
    return int(math.floor((lon + 180.0) / 6.0)) % 60 + 1


def utm_epsg(lat: float, lon: float) -> int:
    """EPSG code of the WGS84 / UTM zone containing (lat, lon).

    326xx on the northern hemisphere, 327xx on the southern one.
    """
    zone = utm_zone(lon)
    return (32600 if lat >= 0.0 else 32700) + zone


class LocalFrame:
    """WGS84 <-> local metric ENU frame centered on (lat, lon), rotated by
    ``rotation_deg`` counter-clockwise about the origin."""

    def __init__(self, lat: float, lon: float, rotation_deg: float = 0.0) -> None:
        self.lat = float(lat)
        self.lon = float(lon)
        self.rotation_deg = float(rotation_deg) % 360.0
        self.epsg = utm_epsg(self.lat, self.lon)
        self._fwd = Transformer.from_crs("EPSG:4326", f"EPSG:{self.epsg}", always_xy=True)
        self._inv = Transformer.from_crs(f"EPSG:{self.epsg}", "EPSG:4326", always_xy=True)
        self.x0, self.y0 = self._fwd.transform(self.lon, self.lat)
        theta = math.radians(self.rotation_deg)
        self._cos = math.cos(theta)
        self._sin = math.sin(theta)

    # -- forward ---------------------------------------------------------
    def to_local(self, lons: Sequence[float], lats: Sequence[float]) -> np.ndarray:
        """Project degree arrays to an (n, 2) float array of local meters.

        Applies the UTM projection, the center offset and the rotation in one
        vectorized pass.
        """
        lon_a = np.asarray(lons, dtype=float)
        lat_a = np.asarray(lats, dtype=float)
        if lon_a.size == 0:
            return np.zeros((0, 2), dtype=float)
        x, y = self._fwd.transform(lon_a, lat_a)
        x = np.asarray(x, dtype=float) - self.x0
        y = np.asarray(y, dtype=float) - self.y0
        if self.rotation_deg != 0.0:
            xr = x * self._cos - y * self._sin
            yr = x * self._sin + y * self._cos
            x, y = xr, yr
        return np.column_stack((x, y))

    def point_to_local(self, lon: float, lat: float) -> tuple[float, float]:
        arr = self.to_local([lon], [lat])
        return float(arr[0, 0]), float(arr[0, 1])

    # -- inverse ---------------------------------------------------------
    def to_wgs84(self, xy: Iterable[Sequence[float]]) -> np.ndarray:
        """Inverse of :meth:`to_local`; returns an (n, 2) array of (lon, lat).

        Used only to compute the Overpass bounding box, never on the SceneGraph
        path: no lat/lon may cross the SceneGraph boundary.
        """
        pts = np.asarray(list(xy), dtype=float).reshape(-1, 2)
        if pts.size == 0:
            return np.zeros((0, 2), dtype=float)
        x, y = pts[:, 0], pts[:, 1]
        if self.rotation_deg != 0.0:
            xu = x * self._cos + y * self._sin
            yu = -x * self._sin + y * self._cos
            x, y = xu, yu
        lon, lat = self._inv.transform(x + self.x0, y + self.y0)
        return np.column_stack((np.asarray(lon, dtype=float), np.asarray(lat, dtype=float)))


def crop_square(radius_m: float) -> Polygon:
    """The axis-aligned crop square of side ``2 * radius_m`` centered on (0,0)."""
    r = float(radius_m)
    return box(-r, -r, r, r)


def _polygonal_parts(geom: BaseGeometry) -> list[Polygon]:
    if geom.is_empty:
        return []
    if isinstance(geom, Polygon):
        return [geom]
    if isinstance(geom, MultiPolygon):
        return [p for p in geom.geoms if not p.is_empty]
    parts: list[Polygon] = []
    for sub in getattr(geom, "geoms", []):
        parts.extend(_polygonal_parts(sub))
    return parts


def clip_polygon(geom: BaseGeometry, square: Polygon, min_area_m2: float = 1.0) -> list[Polygon]:
    """Intersect a polygon with the crop square.

    A half building at the edge becomes a clean cut face (03), and a result
    that falls apart into several pieces is returned as several polygons so the
    caller can emit them as separate SceneGraph entries.
    """
    if geom.is_empty:
        return []
    clipped = geom.intersection(square)
    return [p for p in _polygonal_parts(clipped) if p.area >= min_area_m2]


def clip_line(geom: BaseGeometry, square: Polygon, min_length_m: float = 1.0) -> list[LineString]:
    """Intersect a linestring with the crop square, splitting MultiLineStrings."""
    if geom.is_empty:
        return []
    clipped = geom.intersection(square)
    out: list[LineString] = []
    stack: list[BaseGeometry] = [clipped]
    while stack:
        g = stack.pop()
        if g.is_empty:
            continue
        if isinstance(g, LineString):
            if g.length >= min_length_m and len(g.coords) >= 2:
                out.append(g)
        elif isinstance(g, MultiLineString):
            stack.extend(g.geoms)
        else:
            stack.extend(getattr(g, "geoms", []))
    return out


def in_square(x: float, y: float, radius_m: float) -> bool:
    """Containment test used for point features (trees)."""
    r = float(radius_m)
    return -r <= x <= r and -r <= y <= r

File: apps/web/lib/engine/osm/project.ts
/**
 * Local metric frame: WGS84 -> UTM -> centered ENU, plus rotation and crop.
 * Mirrors `services/bake/app/geom/project.py`: everything here is meters in a
 * local ENU frame with the requested center at (0,0), x east, y north. Web
 * Mercator is never used. See `tmerc.ts` for the projection math and its
 * validation against `pyproj`.
 */
import { area, boxRing, clipConvex, type Point, type Ring } from "./geometry";
import { utmForward, utmInverse, utmZoneFor, type UtmZone } from "./tmerc";

export class LocalFrame {
  readonly lat: number;
  readonly lon: number;
  readonly rotationDeg: number;
  readonly zone: UtmZone;
  private readonly x0: number;
  private readonly y0: number;
  private readonly cos: number;
  private readonly sin: number;

  constructor(lat: number, lon: number, rotationDeg = 0.0) {
    this.lat = lat;
    this.lon = lon;
    this.rotationDeg = ((rotationDeg % 360) + 360) % 360;
    this.zone = utmZoneFor(lat, lon);
    const [x0, y0] = utmForward(lat, lon, this.zone);
    this.x0 = x0;
    this.y0 = y0;
    const theta = (this.rotationDeg * Math.PI) / 180;
    this.cos = Math.cos(theta);
    this.sin = Math.sin(theta);
  }

  get epsg(): number {
    return this.zone.epsg;
  }

  /** Project WGS84 (lon, lat) pairs to local meters, applying the center offset and rotation. */
  toLocal(lonLat: Point[]): Point[] {
    return lonLat.map(([lon, lat]) => this.pointToLocal(lon, lat));
  }

  pointToLocal(lon: number, lat: number): Point {
    const [x, y] = utmForward(lat, lon, this.zone);
    let dx = x - this.x0;
    let dy = y - this.y0;
    if (this.rotationDeg !== 0) {
      const xr = dx * this.cos - dy * this.sin;
      const yr = dx * this.sin + dy * this.cos;
      dx = xr;
      dy = yr;
    }
    return [dx, dy];
  }

  /** Inverse of `toLocal`; used only to compute the Overpass bbox, never past the SceneGraph boundary. */
  toWgs84(xy: Point[]): Point[] {
    return xy.map(([x, y]) => this.pointToWgs84(x, y));
  }

  pointToWgs84(x: number, y: number): Point {
    let ux = x;
    let uy = y;
    if (this.rotationDeg !== 0) {
      ux = x * this.cos + y * this.sin;
      uy = -x * this.sin + y * this.cos;
    }
    const [lat, lon] = utmInverse(ux + this.x0, uy + this.y0, this.zone);
    return [lon, lat];
  }
}

/** The axis-aligned crop square of side `2 * radiusM` centered on (0,0), as a CCW ring. */
export function cropSquare(radiusM: number): Ring {
  const r = radiusM;
  return boxRing(-r, -r, r, r);
}

/** Intersect a polygon ring with the crop square; drops pieces under `minAreaM2`. See geometry.ts's clipConvex caveat. */
export function clipPolygonToSquare(ring: Ring, square: Ring, minAreaM2 = 1.0): Ring[] {
  const clipped = clipConvex(ring, square);
  if (clipped.length < 3) return [];
  return area(clipped) >= minAreaM2 ? [clipped] : [];
}

/** Intersect an open polyline with the crop square (Cohen-Sutherland-free: the square's 4 half-planes, sequentially). */
export function clipLineToSquare(line: Point[], radiusM: number, minLengthM = 1.0): Point[][] {
  let segments: Point[][] = [line];
  const planes: [number, (p: Point) => number][] = [
    [radiusM, (p) => p[0] + radiusM], // x >= -r
    [radiusM, (p) => radiusM - p[0]], // x <= r
    [radiusM, (p) => p[1] + radiusM], // y >= -r
    [radiusM, (p) => radiusM - p[1]], // y <= r
  ];
  for (const [, inside] of planes) {
    const next: Point[][] = [];
    for (const seg of segments) next.push(...clipPolylineHalfPlane(seg, inside));
    segments = next;
  }
  return segments
    .filter((s) => s.length >= 2)
    .filter((s) => polylineLength(s) >= minLengthM);
}

function polylineLength(pts: Point[]): number {
  let len = 0;
  for (let i = 1; i < pts.length; i++) {
    const dx = pts[i][0] - pts[i - 1][0];
    const dy = pts[i][1] - pts[i - 1][1];
    len += Math.hypot(dx, dy);
  }
  return len;
}

function clipPolylineHalfPlane(pts: Point[], inside: (p: Point) => number): Point[][] {
  const out: Point[][] = [];
  let current: Point[] = [];
  for (let i = 0; i < pts.length; i++) {
    const p = pts[i];
    const pIn = inside(p) >= -1e-9;
    if (i > 0) {
      const prev = pts[i - 1];
      const prevIn = inside(prev) >= -1e-9;
      if (prevIn !== pIn) {
        const t = inside(prev) / (inside(prev) - inside(p));
        const ix = prev[0] + t * (p[0] - prev[0]);
        const iy = prev[1] + t * (p[1] - prev[1]);
        current.push([ix, iy]);
        if (!pIn) {
          out.push(current);
          current = [];
        }
      }
    }
    if (pIn) current.push(p);
  }
  if (current.length >= 2) out.push(current);
  return out;
}

/** Containment test for point features (trees), matching `project.py:in_square`. */
export function inSquare(x: number, y: number, radiusM: number): boolean {
  return -radiusM <= x && x <= radiusM && -radiusM <= y && y <= radiusM;
}
